c2corg_api/models/cache_version.py

```diff
@@ -0,0 +1,354 @@
+import logging
+
+from c2corg_api.models import Base, schema, DBSession
+from c2corg_api.models.document import Document
+from c2corg_api.models.outing import OUTING_TYPE
+from c2corg_api.models.route import ROUTE_TYPE
+from c2corg_api.models.waypoint import WAYPOINT_TYPE
+from sqlalchemy.orm import relationship
+from sqlalchemy.sql.schema import Column, ForeignKey
+from sqlalchemy.sql.sqltypes import Integer
+from sqlalchemy import event, DDL, text
+
+log = logging.getLogger(__name__)
+
+
+class CacheVersion(Base):
+    """This table contains a version number for each document that reflects
+    the state of the document and all its associated documents.
+    When the document or one of its associated documents is updated, the
+    version of the document is incremented, so that outdated documents are not
+    served from the cache.
+    """
+    __tablename__ = 'cache_versions'
+
+    document_id = Column(
+        Integer, ForeignKey(schema + '.documents.document_id'),
+        primary_key=True)
+    document = relationship(
+        Document, primaryjoin=document_id == Document.document_id
+    )
+
+    version = Column(Integer, nullable=False, server_default='1')
+
+
+# Trigger that creates a new entry in the `CacheVersion` table when a new
+# document is created.
+trigger_ddl = DDL("""
+CREATE OR REPLACE FUNCTION guidebook.create_cache_version() RETURNS TRIGGER AS
+$BODY$
+BEGIN
+    INSERT INTO
+        guidebook.cache_versions(document_id)
+        VALUES(new.document_id);
+    RETURN null;
+END;
+$BODY$
+language plpgsql;
+
+CREATE TRIGGER guidebook_documents_insert
+AFTER INSERT ON guidebook.documents
+FOR EACH ROW
+EXECUTE PROCEDURE guidebook.create_cache_version();
+""")
+event.listen(Document.__table__, 'after_create', trigger_ddl)
+
+
+# functions to update the cache version if a document or associations
+# have changed
+update_cache_version_ddl = DDL("""
+CREATE OR REPLACE FUNCTION guidebook.update_cache_version(p_document_id integer, p_document_type character varying(1)) --  # noqa: E501
+  RETURNS void AS
+$BODY$
+BEGIN
+  -- function to update all dependent documents if a document changes
+
+  -- update the version of the document itself
+  PERFORM guidebook.increment_cache_version(p_document_id);
+
+  -- update the version of linked documents (direct associations)
+  PERFORM guidebook.update_cache_version_of_linked_documents(p_document_id);
+
+  if p_document_type = 'w' then
```

Comment:
What about other document types to come (images, articles, profiles, books, accident reports)?
Will they required custom ``update_cache_version_*`` functions?